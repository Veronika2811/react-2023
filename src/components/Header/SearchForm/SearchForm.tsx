import { useRouter } from 'next/router';
import { FormEvent, useRef } from 'react';

import Button from '../../UI/button/Button';
import { DEFAULT_QUERY_PARAMS } from '@/utils/constants/constants';

import classes from './SearchForm.module.css';

const SearchForm = () => {
  const router = useRouter();
  const { name, perPage } = router.query;

  const inputRef = useRef<HTMLInputElement | null>(null);

  const handleSubmit = (e: FormEvent) => {
    e.preventDefault();
    if (inputRef.current) {
      const currentValue = inputRef.current.value.trim();

      router.push({
        pathname: '/',
        query: {
          ...(currentValue ? { name: currentValue } : {}),
          page: DEFAULT_QUERY_PARAMS.currentPage,
          perPage: perPage || DEFAULT_QUERY_PARAMS.perPage,
        },
      });
    }
  };

  return (
    <form className={classes.input} onSubmit={handleSubmit}>
      <input
        type="text"
        name="search"
        defaultValue={name}
        placeholder="Enter request"
        className={classes.input__box}
        autoFocus
        ref={inputRef}
        data-testid="search-input"
      />
      <Button type="submit" data-testid="search-button">
        Search
      </Button>
    </form>
  );
};

export default SearchForm;
