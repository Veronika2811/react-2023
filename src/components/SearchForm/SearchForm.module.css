.input {
  display: flex;
  flex-wrap: wrap;
  gap: 20px;
}

.input__box {
  padding: 0.6em 1.2em;
  font-size: 1em;
  border: none;
  border-radius: 5px;
}

.input__box:focus {
  outline: 1px solid #4ac7bc;
}