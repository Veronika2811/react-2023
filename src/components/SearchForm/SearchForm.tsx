import { FormEvent, useRef } from 'react';
// import { useSearchParams } from 'react-router-dom';

import Button from '../UI/button/Button';
import { useAppSelector, useAppDispatch } from '@/store/hooks';
import { charactersSaveSearchQuery } from '@/store/slice/charactersSlice';
import { RootState } from '@/store/store';
import {
  // DEFAULT_PAGE,
  LOCAL_STORAGE_KEY,
  // PAGE_URL_PARAMETER_KEY,
} from '@/constants/constants';

import classes from './SearchForm.module.css';

const SearchForm = () => {
  // const [, setSearchParams] = useSearchParams();

  const query = useAppSelector(
    (state: RootState) => state.CHARACTERS_SLICE.query
  );
  const dispatch = useAppDispatch();

  const inputRef = useRef<HTMLInputElement | null>(null);

  const handleSubmit = (e: FormEvent) => {
    e.preventDefault();
    if (inputRef.current) {
      const currentValue = inputRef.current.value.trim();
      localStorage.setItem(LOCAL_STORAGE_KEY, currentValue);

      dispatch(charactersSaveSearchQuery(currentValue));

      // setSearchParams((searchParams) => {
      //   searchParams.set(PAGE_URL_PARAMETER_KEY, DEFAULT_PAGE);
      //   return searchParams;
      // });
    }
  };

  return (
    <form className={classes.input} onSubmit={handleSubmit}>
      <input
        type="text"
        name="search"
        defaultValue={query}
        placeholder="Enter request"
        className={classes.input__box}
        autoFocus
        ref={inputRef}
        data-testid="search-input"
      />
      <Button type="submit" data-testid="search-button">
        Search
      </Button>
    </form>
  );
};

export default SearchForm;
