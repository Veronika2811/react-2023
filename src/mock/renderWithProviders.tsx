import React, { PropsWithChildren } from 'react';
import { Provider } from 'react-redux';
// import { HashRouter } from 'react-router-dom';
import type { PreloadedState } from '@reduxjs/toolkit';
import { render } from '@testing-library/react';
import type { RenderOptions } from '@testing-library/react';

import { AppStore, RootState, setupStore } from '@/store/store';

interface ExtendedRenderOptions extends Omit<RenderOptions, 'queries'> {
  preloadedState?: PreloadedState<RootState>;
  store?: AppStore;
}

export const renderWithProviders = (
  ui: React.ReactElement,
  {
    preloadedState = {},
    store = setupStore(preloadedState),
    ...renderOptions
  }: ExtendedRenderOptions = {}
) => {
  const Wrapper = ({ children }: PropsWithChildren): JSX.Element => {
    return (
      // <HashRouter>
        <Provider store={store}>{children}</Provider>
      // </HashRouter>
    );
  };
  return { store, ...render(ui, { wrapper: Wrapper, ...renderOptions }) };
};

export default renderWithProviders;
