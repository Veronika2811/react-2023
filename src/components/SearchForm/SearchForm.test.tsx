import { expect, it } from 'vitest';
import { screen, fireEvent, waitFor } from '@testing-library/react';
import '@testing-library/jest-dom';

import SearchForm from './SearchForm';
import renderWithProviders from '@/mock/renderWithProviders';

describe('SearchForm component', () => {
  const MOCK_KEY = 'Veronika2811-react-2023__searchRequest';
  const MOCK_VALUE = 'test-value';

  it('renders correctly SearchForm component', () => {
    const container = renderWithProviders(<SearchForm />);
    expect(container).toMatchSnapshot();
  });

  it('should save the entered value in local storage', () => {
    renderWithProviders(<SearchForm />);

    const searchInput = screen.getByTestId('search-input');
    const searchButton = screen.getByTestId('search-button');

    fireEvent.change(searchInput, { target: { value: MOCK_VALUE } });
    fireEvent.click(searchButton);

    expect(localStorage.getItem(MOCK_KEY)).toBe(MOCK_VALUE);
  });

  it('should get the value from local storage', () => {
    renderWithProviders(<SearchForm />);

    const searchInputValue = (
      screen.getByTestId('search-input') as HTMLInputElement
    ).value;

    waitFor(() => expect(searchInputValue).toBe(MOCK_VALUE));
  });
});
