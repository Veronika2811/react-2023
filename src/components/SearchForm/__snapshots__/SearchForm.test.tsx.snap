// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`SearchForm component > renders correctly SearchForm component 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <form
        class="_input_49e12d"
      >
        <input
          class="_input__box_49e12d"
          data-testid="search-input"
          name="search"
          placeholder="Enter request"
          type="text"
          value=""
        />
        <button
          class="_button_634e7c"
          data-testid="search-button"
          type="submit"
        >
          Search
        </button>
      </form>
    </div>
  </body>,
  "container": <div>
    <form
      class="_input_49e12d"
    >
      <input
        class="_input__box_49e12d"
        data-testid="search-input"
        name="search"
        placeholder="Enter request"
        type="text"
        value=""
      />
      <button
        class="_button_634e7c"
        data-testid="search-button"
        type="submit"
      >
        Search
      </button>
    </form>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "store": {
    "@@observable": [Function],
    "dispatch": [Function],
    "getState": [Function],
    "replaceReducer": [Function],
    "subscribe": [Function],
  },
  "unmount": [Function],
}
`;
